### PR TITLE
Update: Add a fixer for `dot-location`

### PR DESCRIPTION
**Pull request by not-an-aardvark**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://github.com/eslint/eslint/blob/master/templates/bug-report.md))
[ ] New rule ([template](https://github.com/eslint/eslint/blob/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md))
[x] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**Please check each item to ensure your pull request is ready:**

- [x] I've read the [pull request guide](http://eslint.org/docs/developer-guide/contributing/pull-requests)
- [x] I've included tests for my change
- [x] I've updated documentation for my change (if appropriate)

**What changes did you make? (Give an overview)**

This adds an autofixer for [`dot-location`](http://eslint.org/docs/rules/dot-location).

Given a `MemberExpression` where the dot is in the wrong location, the fixer will move the dot to the correct location. All comments between the object and the property are preserved.

```js
/* eslint dot-location: [2, "property"] */

foo.
  bar;

baz. /* comment */
  qux;

// gets fixed to:

foo
  .bar;

baz /* comment */
  .qux;
```

```js
/* eslint dot-location: [2, "object"] */

foo
  .bar;

5
  .toExponential();

// gets fixed to:

foo.
  bar;

5 .
  toExponential();

// (A space is added after `5` to ensure that the dot isn't parsed as a decimal point)
```

**Is there anything you'd like reviewers to focus on?**

I marked this as `fixable: "code"`, but should it be `fixable: "whitespace"` instead since it doesn't change the AST?

### CODE REVIEW

**Comment by platinumazure** [comment]
I think fixable: code is correct since you are not just adding or removing whitespace, but tokens as well.

**Comment by gcochard** [comment]
There's [currently no functional difference](https://github.com/eslint/eslint/issues/6010#issuecomment-216401044) between `code` and `whitespace`, but FWIW I'd go with whitespace as you're not modifying the AST.

**Comment by platinumazure** [comment]
We've used `fixable: "code"` when adding or removing parentheses (including in no-extra-parens, where we knew the AST would be unmodified). Maybe someone else on the team will explain it better, but to me, as soon as you add/remove/move a token, even for an equivalent AST, it's a code change.

**Comment by not-an-aardvark** [comment]
The reasoning in this case would be that we're not actually changing any *tokens*, either; we're basically changing the amount of whitespace that appears before and after the dot.

(That said, we are also moving comments around, so it's not just whitespace.)

**Comment by mysticatea** [comment]
Thank you for this PR!

I will be a champion for this change.

- I think good that `dot-location` also has autofix since very similar [comma-style](http://eslint.org/docs/rules/comma-style) has autofix.

By the way, `comma-style` is `fixable: "code"`.

File: lib/rules/dot-location.js [hunk 1]
```diff
@@ -44,14 +46,30 @@ module.exports = {
...
         function checkDotLocation(obj, prop, node) {
             const dot = sourceCode.getTokenBefore(prop);
+            const textBeforeDot = sourceCode.getText().slice(obj.range[1], dot.range[0]);
+            const textAfterDot = sourceCode.getText().slice(dot.range[1], prop.range[0]);
 
             if (dot.type === "Punctuator" && dot.value === ".") {
                 if (onObject) {
                     if (!astUtils.isTokenOnSameLine(obj, dot)) {
-                        context.report(node, dot.loc.start, "Expected dot to be on same line as object.");
+                        const neededTextAfterObj = obj.type === "Literal" && typeof obj.value === "number" && /^\d*$/.test(obj.raw)
```
Comment (mysticatea) [inline lib/rules/dot-location.js]: A note to me: `/^\d*$/` will be replaced by `astUtils.isDecimalInteger` after #7185 is merged.

Comment (not-an-aardvark) [inline lib/rules/dot-location.js]: #7185 is merged now, so I rebased and switched to `astUtils.isDecimalInteger`.

**Comment by vitorbal** [comment]
I'll support this! :+1:

**Pull request merged**